File: fptest.js

function threshold(tValue) {
	const k = 0x100_0000_0000_0000  // 2^56

	if (tValue < 1.0) {
		return Math.floor(k*tValue + 0.5)
	}
	return Math.floor(k/tValue + 0.5)
}

function samplingRate(tValue) {
	if (tValue < 1.0) {
		return Math.floor(1.0/tValue + 0.5)
	}
	return Math.floor(tValue + 0.5)
}

function samplingProbability(tValue) {
	if (tValue < 1.0) {
		return tValue
	}
	return 1.0 / tValue
}

function all(args) {
    for (var i=0; i<args.length; i++) {
        var arg = process.argv[i]
        var tValue = parseFloat(arg)
        if (isNaN(tValue)) {
            console.log("bad value: %s", arg)
            continue
        }

		console.log("tValue: %f", tValue)
		console.log("threshold: %d", threshold(tValue))
		console.log("samplingRate: %d", samplingRate(tValue))
		console.log("samplingProbability: %f", samplingProbability(tValue))
		console.log()
	}
}

function thresh(arg) {
	var tValue = parseFloat(arg)
	if (isNaN(tValue)) {
		return "bad value: " + arg
	}
	return ""+threshold(tValue)
}

function rate(arg) {
	var tValue = parseFloat(arg)
	if (isNaN(tValue)) {
		return "bad value: " + arg
	}
	return ""+samplingRate(tValue)
}

function prob(arg) {
	var tValue = parseFloat(arg)
	if (isNaN(tValue)) {
		return "bad value: " + arg
	}
	return samplingProbability(tValue).toFixed(14)
}

function main() {
	var args = process.argv.slice(2)
	if (args.length == 0) {
		console.log("usage: node fptest.js command <tValue> ...")
		return
	}
	if (args[0] == "all") {
		all(args)
	} else if (args[0] == "thresh") {
		console.log(thresh(args[1]))
	} else if (args[0] == "rate") {
		console.log(rate(args[1]))
	}
	else if (args[0] == "prob") {
		console.log(prob(args[1]))
	} else {
		console.log("bad command: %s", args[0])
	}
}

main()

File: fptest.py
#!/usr/bin/env python3

import sys
import os
import random

def threshold(tValue):
    k = 0x100_0000_0000_0000  # 2^56
    if tValue < 1.0:
        return int(k * tValue + 0.5)
    else:
        return int(k / tValue + 0.5)

def samplingRate(tValue):
    if tValue < 1.0:
        return int(1.0/tValue + 0.5)
    else:
        return int(tValue + 0.5)

def samplingProbability(tValue):
    if tValue < 1.0:
        return tValue
    else:
        return 1.0/tValue

def all(args):
    for i in args:
        tValue = float(i)

        print(f"tValue: {tValue:.14g}")
        print(f"threshold: {threshold(tValue):d}")
        print(f"samplingRate: {samplingRate(tValue):d}")
        print(f"samplingProbability: {samplingProbability(tValue):.14g}")
        print()

def thresh(arg):
    tValue = float(arg)
    return f"{threshold(tValue):d}"

def rate(arg):
    tValue = float(arg)
    return f"{samplingRate(tValue):d}"

def prob(arg):
    tValue = float(arg)
    return f"{samplingProbability(tValue):.14g}"


if __name__ == '__main__':
    if len(sys.argv) == 1:
        print("Usage: fptest.py command <tValue> ...")
        print("command: all, thresh, rate, prob")
        sys.exit(1)

    args = sys.argv[2:]
    command = sys.argv[1]
    if command == "all":
        all(args)
    elif command == "thresh":
        print(thresh(args[0]))
    elif command == "rate":
        print(rate(args[0]))
    elif command == "prob":
        print(prob(args[0]))
    else:
        print("Usage: fptest.py command <tValue> ...")
        print("command: all, thresh, rate, prob")
        sys.exit(1)

